Step3: GCD 작업 스케줄링

TTOzzi opened the pull request:
* 코드로 NavigationController를 embed 해주기 위해 AppDelegate 에서 doodleViewController 와 NavigationController 를 생성해주었습니다. 기존에 스토리보드에서 만든 NavigationController 를 사용하지 않았는데 이렇게 하는 게 맞나요??
* JSON 데이터 디코드, 이미지 다운로드, 셀에 이미지 설정 모두를 비동기로 동작하게 해주었는데 DispatchQueue 를 아직 정확하게 이해하지 못해 제가 한 방법이 제대로 된 구현이 맞는지 모르겠습니다.
* DoodleViewController 에서 DataSource 에 해당하는 부분을 분리해줄까 고민하였는데 그러면 UICollectionViewController 를 상속받은 의미가 없는 것 같아 분리하지 않았습니다. 그러니 DoodleViewController 의 역할이 너무 커진 것 같습니다. 이럴 땐 분리해주어야 할까요?
* 앱 전체에서 하나의 PhotoDataManager 를 써주도록 하기 위해 AppDelegate 에 인스턴스를 만들어 준 후 다른 클래스들에서 AppDelegate 를 통해 PhotoDataManager 에 접근하도록 해주었습니다. 이런 방법을 써도 될까요?
* ~~이미지를 다운 받을 때마다 콜렉션뷰를 리로드해주니 이미지를 다운받는 동안은 Save 버튼이 제대로 작동하지 않았습니다. 이를 해결해주기 위해 화면 표시에 당장 필요한 이미지 40개를 다운 받을 동안만 리로드 해주고 그 이후엔 이미지를 모두 받았을 때 리로드 해주도록 해주었습니다. 이런 방법을 써도 되는 건가요?~~
* 다른 멤버들과 이야기하다 비동기로 동작하는 메소드라도 for문에 넣으면 동기방식으로 동작하게 된다는 것을 들었습니다. 그리고 한 번에 이미지를 받아오는 로직이 지금은 100개 정도라 괜찮지만, 사진이 많아진다면 비효율적일 것 같아 셀 생성 시점에 필요한 이미지를 로드해주도록 수정하였습니다.

TTOzzi committed:
NSNotification.Name extension 추가, 뷰, 모델 폴더별로 정리

TTOzzi committed:
guard 문 처리방식 변경

Notification 에서 받아온 데이터를 한번에 처리해주도록 수정

TTOzzi committed:
NavigationBar 에 BarButton 추가

System Item 을 Add 로 변경해줌

TTOzzi committed:
JSON 데이터 처리 객체 추가

URL 에서 데이터를 가져와 DoodleImage 구조체로 만들어 DoodleImages 배열에 넣어줌

TTOzzi committed:
DoodleViewController 생성

UICollectionViweController 를 상속받음

TTOzzi committed:
add 버튼 추가

TTOzzi committed:
DoodleViewController Embed

AppDelegate 에서 앱 실행 시 NavigationController 에 DoodleViewController 를 embed 하도록 해줌

TTOzzi committed:
add 버튼 클릭 시 DoodleViewController present 구현

TTOzzi committed:
DoodleCollectionView 배경색, Title 지정

배경색을 어두운 회색으로, Title 을 "Doodles" 로 지정해줌

TTOzzi committed:
Close 버튼 구현

Close 버튼 클릭 시 dismiss 하도록 구현

TTOzzi committed:
DoodleView 커스텀 셀 구현

TTOzzi committed:
DoodleView 셀 크기 110 x 50 으로 설정

UICollectionViewDelegateFlowLayout 프로토콜 채택 후 사이즈 설정

TTOzzi committed:
View Controller BarButton 위치 변경

TTOzzi committed:
DoodleViewCell 에 ImageView 추가

이미지뷰가 셀에 꽉 차게 출력되도록 오토레이아웃 설정

TTOzzi committed:
index 를 받아 그에 맞는 이미지를 반환하는 함수 구현

TTOzzi committed:
GCD 큐 활용해 셀 이미지 다운로드 및 표시 구현

먼저 JSON 데이터를 받아와 디코드 한 뒤 배열에 넣음.
비동기로 데이터를 받아오기위해 URLSession 의 dataTask 메소드 활용.
디코드한 JSON 데이터의 URL 에서 이미지를 받아옴과 동시에 콜렉션뷰의 셀 이미지를 설정해줌.
completion 을 활용해 하나의 URL 에서 이미지를 받아온 후 이미지를 보내줌.
받아온 이미지를 cellImages 배열에 넣고 콜렉션뷰를 리로드해줌.
리로드 할 때 콜렉션뷰에서는 해당 indexPath 의 이미지가 배열에 존재하면 그 이미지를 넣어주고 존재하지않으면 기본 이미지를 넣어줌.

TTOzzi committed:
JSON decode 도 비동기로 동작하도록 변경

JSON decode 하는데도 시간이 걸려 콜렉션 뷰가 늦게 나타나는 문제 수정
decode 한 후 completion 으로 JSON 데이터 배열을 넘기도록 수정

TTOzzi committed:
셀에 LongPress 액션 추가

UILongPressGestureRecognizer 를 활용해 셀이 0.5초 이상 눌리면 cellPressed 메소드가 호출되도록 해줌

TTOzzi committed:
셀 LongPress 시 Save UIMenu 를 표시하도록 구현

Save 메뉴 클릭시 saveItemTabbed() 메소드가 호출 됨
셀이 FirstResponder 가 되게 해주기 위해 셀의 canBecomFirstResponder 프로퍼티를 override 해주고 선택된 셀의 becomeFirstResponder 메소드 호출

TTOzzi committed:
Save 메뉴 클릭 시 사진 저장 구현

AppDelegate 에 PhotoDataManager 를 두어 어떤 곳이든 하나의 PhotoDataManager 를 사용하도록 구현
Save 메뉴를 클릭하면 셀의 이미지를 selectedImage 에 넣어주고 그 이미지를 PHPhotoLibrary 를 활용해 사진앱에 추가

TTOzzi committed:
노티피케이션 수정

userInfo 의 키값을 모두 같은 값으로 보내주니 인식을 못하는 문제 발생
다시 각각 다른 키값을 만들어줌

TTOzzi committed:
접근제어자 설정, 코드 정리

TTOzzi committed:
CollectionView 이미지 로드 로직 변경

당장 표시하기위해 필요한 이미지 40개 까지는 이미지를 받을 때 마다 리로드 해주고 그 이후엔 모두 다운받았을때 리로드 해주도록 수정

godrm commented on PhotosApp/PhotosApp/AppDelegate.swift:
```diff
@@ -10,7 +10,14 @@ import UIKit
+    
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+        doodleViewController = DoodleViewController(collectionViewLayout: UICollectionViewFlowLayout())
+        doodleNavigationController = UINavigationController(rootViewController: doodleViewController)
```
일부러 코드로 한게 아니라면 스토리보드 상에서 처리해도 됩니다. 
이렇게하면 원래 있던 ViewController 어디갔죠? 화면 흐름이 이어져야 하는데 DoodleViewController만 보이지 않나요?

TTOzzi replied:
이렇게 해도 화면이 잘 이어져서 이렇게 하는게 맞는 방법인줄 알았습니다.
기존 스토리보드의 NavigationController 에 embed 해야하나요?
그렇다면 스토리보드에 만들어줬던 NavigationController 를 코드로 가져오는 방법이 있을까요..?

godrm replied:
찾아보세요. 쉬운 방법을 돌아서 어렵게 하고 있으니까요

TTOzzi replied:
넵
다른 방법으로 다시 구현해보겠습니다
감사합니다!

TTOzzi committed:
비동기 이미지 로드 로직 변경

기존 이미지를 한번에 다 받아오는 방식에서 셀 제작에 필요한 이미지만 받아오도록 수정
한번 받아온 이미지는 다시 받아오지 않고 재사용하도록 구현

godrm commented on PhotosApp/PhotosApp/DoodleViewController.swift:
```diff
@@ -0,0 +1,104 @@
+        decoder.decodeJson { (images) in
+            guard let images = images as? [DataDecoder.DoodleImage] else { return }
+            self.fetchImages(images)
+        }
```
여기서 decoder에서 이미지를 처리하는 코드를 ViewController가 하는게 어색합니다.
ViewController가 디코딩해서, 이미지 캐싱, 셀 생성까지 모두 포함하면서 섞여있네요. 분리해보세요.
cellImages에서 이미지를 배열로 갖고 있는 것도 좋은 방법은 아닙니다. 이미지는 용량이 커서 개수가 많아지면 앱 메모리를 많이 차지하고 결국에 메모리가 부족해서 죽을 수도 있습니다.

TTOzzi replied:
이 부분도 PR 을 보낸 상태로 수정해서 반영이 안되었네요.
셀을 만들 때 마다 이미지를 다운받아오는 방식이 비효율적인것같아 한번 받아온 이미지는 cellImages 배열이 갖고있게 해주었는데 만들 때 마다 이미지를 받아오는게 더 나은건가요??

godrm replied:
아뇨. 그러니까 메모리에 모든걸 저장하는 게 아니라 이미지처럼 용량이 큰건 디스크에 캐싱을 해야합니다. 
받을걸 또 받아올 필요는 없으니까요.

godrm commented on PhotosApp/PhotosApp/DoodleViewController.swift:
```diff
@@ -0,0 +1,104 @@
+                cell.setImage(self.cellImages[indexPath.item])
+            } else {
+                cell.setImage(self.defaultImage)
+            }
```
이미 콜랙션뷰 처리하는 델리게이트는 main에서 실행됩니다. `DispatchQueue.main.async`는 의미 없는 표현이네요.

TTOzzi replied:
앗 이 부분은 방금 로직을 다르게 수정했어요
PR 을 보내놓고 수정해서 반영이 안됐네요

godrm replied:
그럼 해결된거겠죠?

godrm commented on PhotosApp/PhotosApp/DoodleViewController.swift:
```diff
@@ -0,0 +1,104 @@
+                    }
+                }
+            }
+        }
```
이미지를 가져오는 것도 여기서 다 구현하는 게 아니라 별도 객체로 분리하세요. 
앞 단계에서 PH 프레임워크를 사용해봤는데, 비슷한 방식인데 처리하는 게 너무 다르지 않나요?

TTOzzi replied:
분리해보도록 하겠습니다

godrm commented on PhotosApp/PhotosApp/DoodleViewController.swift:
```diff
@@ -0,0 +1,104 @@
+    }
+    
+    @objc func saveItemTabbed() {
+        app.photoDataManager.addImage(image: selectedImage)
```
이걸 하려고 AppDelegate까지 참조를 하고 있었군요. 
꼭 AppDelegate가 소유하는 객체여야 할까요? 꼭 의존하고 필요한거라면, 의존성 주입을 고려해보는 게 좋았을 것 같습니다. 이 부분도 공부해보세요.

godrm commented on PhotosApp/PhotosApp/DoodleViewController.swift:
```diff
@@ -0,0 +1,104 @@
+
+extension DoodleViewController: UICollectionViewDelegateFlowLayout {
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+        return CGSize(width: 110, height: 50)
```
이전 단계에서 했던 피드백을 떠올려보세요. 하드코딩하지마세요.

godrm commented on PhotosApp/PhotosApp/Extension/NSNotificationNameExtension.swift:
```diff
@@ -0,0 +1,17 @@
+    static let photoRemoved = NSNotification.Name("photoRemoved")
+    static let photoInserted = NSNotification.Name("photoInserted")
+    static let photoChanged = NSNotification.Name("photoChanged")
+    static let photoEnumerated = NSNotification.Name("photoEnumerated")
```
여기 있는 노티들은 어떤 객체를 위한 걸까요? 
Cell Identifier처럼 각 객체와 관련된 값은 해당 객체에 선언하세요.

TTOzzi replied:
저번 피드백에서 NSNotification.Name 을 확장하는 방법도 있다고 하셔서 이렇게 했었는데 이런 방법이 아닌가요?

godrm replied:
확장하지만 다 모아놓는게 아니라 필요한 객체 아래에 포함하라는 겁니다.

godrm commented on PhotosApp/PhotosApp/Model/DataDecoder.swift:
```diff
@@ -0,0 +1,40 @@
+            } catch {
+                print("decode failed: ", error)
+            }
+        }
```
무조건 main queue에서 처리하고 있는 것 같습니다. 특히 네트워크 요청은 메인에서 처리하면 안되고 백그라운드에서 처리하세요.

godrm commented on PhotosApp/PhotosApp/Model/DataDecoder.swift:
```diff
@@ -0,0 +1,40 @@
+    }
+    
+    func loadImage(url: URL, completion: @escaping (_: Any) -> ()) {
+        URLSession.shared.dataTask(with: url) { (data, response, error) in
```
DataDecoder 역할은 어디까지 인가요? 이름은 디코딩만 하는 것 같은데, 이미지 다운로드까지 처리하는 건가요?

TTOzzi replied:
이 부분도 분리해보겠습니다

godrm commented on PhotosApp/PhotosApp/Model/DataDecoder.swift:
```diff
@@ -0,0 +1,40 @@
+                self.doodleImages = try JSONDecoder().decode([DoodleImage].self, from: jsonData)
+                completion(self.doodleImages)
+            } catch {
+                print("decode failed: ", error)
```
try-catch 를 처리할 때는 do 영역과 catch 영역의 흐름을 꼭 분리하세요. 
정상적인 흐름과 예외적인 흐름이 같으면 안됩니다. 사용자가 인지할 수 있는 수준으로 고민해봐야 합니다. 
특히 앞으로 print()를 넣지 마세요.

godrm commented on PhotosApp/PhotosApp/Model/PhotoCollectionViewDataSource.swift:
```diff
@@ -9,15 +9,15 @@
     
     func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
-        return photoDataManager.photoCount
+        return app.photoDataManager.photoCount
```
앱에 하나만 있어야 한다고 하더라도 AppDelegate에 선언하고 여기저기 AppDelegate를 참조해서 쓰는 구조는 좋지 않습니다.

godrm reviewed (changes requested):
다음 피드백을 확인해보세요.

TTOzzi committed:
이미지 데이터 디코드 실패시 기본 이미지 반환 구현

데이터 디코드 실패시 기본 photo 시스템 이미지를 complition 으로 넘겨줌
complition 으로 넘기는 인자를 Any 에서 UIImage 로 변경

TTOzzi committed:
DoodleViewController 에서 DataSource 부분 분리

DataSource 생성시점에서 JSON 디코드가 완료되면 completion 을 통해 Notification 을 보내도록 구현
DoodleViewController 에서 Notification 을 받아 콜렉션 뷰를 리로드해줌

TTOzzi committed:
JSON 다운로드를 백그라운드에서 처리하도록 수정

TTOzzi committed:
이미지 로드하는 객체 분리

TTOzzi committed:
JSON 디코드하는 객체 이름변경 및 구조 수정

TTOzzi committed:
DoodleViewController 코드 역할별로 정리, 하드코딩 수정

TTOzzi committed:
PhotoDataManager 접근 방식 변경

여러곳에서 하나의 PhotoDataManager 를 사용하기위해 클래스 상수를 만들어줌

TTOzzi committed:
Notification Name 선언 위치 변경

TTOzzi committed:
Notification 수정

잘못된 노티를 보내는 문제 수정
